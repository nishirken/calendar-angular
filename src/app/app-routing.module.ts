import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { SigninFormComponent } from './auth/signin-form/signin-form.component';
import { SignupFormComponent } from './auth/signup-form/signup-form.component';
import { WeekComponent } from './week/week/week.component';

export enum RoutePath {
  Signin = 'auth/signin',
  Signup = 'auth/signup',
  Week = 'week',
}

const routes: Routes = [
  { path: '', redirectTo: RoutePath.Signin, pathMatch: 'full' }, // initial
  { path: RoutePath.Signin, component: SigninFormComponent },
  { path: RoutePath.Signup, component: SignupFormComponent },
  { path: RoutePath.Week, component: WeekComponent },
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule { }
