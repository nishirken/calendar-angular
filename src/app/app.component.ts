import { Component, OnDestroy, OnInit } from '@angular/core';
import { Event, NavigationEnd, Router } from '@angular/router';
import { filter, Subscription, withLatestFrom } from 'rxjs';
import { render } from 'calendar-header';
import { Bridge } from 'calendar-header';
import { DateTimeService } from './services/date-time/date-time.service';
import { DateInterval } from '../types';
import { addWeeks, subWeeks } from 'date-fns';
import { routePaths } from './routing/routes-paths';

const isNavEnd = (ev: Event): ev is NavigationEnd =>
  ev instanceof NavigationEnd;

@Component({
  selector: 'app-root',
  template: `<router-outlet></router-outlet>`,
  styleUrls: ['./app.component.sass'],
})
export class AppComponent implements OnInit, OnDestroy {
  private isRendered = false;
  private headerComponent: ReturnType<typeof render> | null = null;
  private headerElement: HTMLElement | null = null;
  private routerSub: Subscription | null = null;
  private weekIntervalSub: Subscription | null = null;
  private weekInterval: DateInterval = { start: new Date(), end: new Date() };
  private bridge: Bridge = {
    onTodayClick: () => this.dateTimeService.selectedDate$.next(new Date()),
    onArrowLeftClick: () => {
      this.router.navigateByUrl(
        routePaths.week(subWeeks(this.dateTimeService.selectedDate$.value, 1))
      );
    },
    onArrowRightClick: () => {
      this.router.navigateByUrl(
        routePaths.week(addWeeks(this.dateTimeService.selectedDate$.value, 1))
      );
    },
    interval: this.weekInterval,
  };

  constructor(
    private readonly router: Router,
    private readonly dateTimeService: DateTimeService
  ) {}

  ngOnInit(): void {
    const el = document.getElementById('header');

    if (!el) {
      throw new Error("#header el hasn't been found");
    }

    this.headerElement = el;
    this.headerElement.style.display = 'none';
    this.headerComponent = render('#header', this.bridge);

    this.weekIntervalSub = this.dateTimeService.selectedWeekInterval$.subscribe(
      (interval) => {
        if (this.bridge.interval) {
          this.headerComponent?.$set({ bridge: { ...this.bridge, interval } });
        }
      }
    );

    this.routerSub = this.router.events
      .pipe(
        filter(isNavEnd),
        withLatestFrom(this.dateTimeService.selectedWeekInterval$)
      )
      .subscribe(([event, interval]) => {
        if (event.urlAfterRedirects.startsWith('/auth')) {
          this.isRendered = false;
          if (this.headerElement) {
            this.headerElement.style.display = 'none';
          }
          return;
        }

        if (this.isRendered) {
          return;
        } else if (this.bridge.interval) {
          this.isRendered = true;
          this.bridge.interval = interval;
          if (this.headerElement) {
            this.headerElement.style.display = 'initial';
          }
        }
      });
  }

  ngOnDestroy(): void {
    this.routerSub?.unsubscribe();
    this.weekIntervalSub?.unsubscribe();
  }
}
