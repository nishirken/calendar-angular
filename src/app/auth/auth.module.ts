import { CUSTOM_ELEMENTS_SCHEMA, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AppRoutingModule } from '../routing/app-routing.module';
import { SigninFormComponent } from './signin-form/signin-form.component';
import { SignupFormComponent } from './signup-form/signup-form.component';
import { AuthApiService } from '../services/auth-api/auth-api.service';
import { ComponentsModule } from '../components/components.module';

@NgModule({
  declarations: [SigninFormComponent, SignupFormComponent],
  imports: [CommonModule, AppRoutingModule, ComponentsModule],
  providers: [AuthApiService],
  schemas: [CUSTOM_ELEMENTS_SCHEMA],
})
export class AuthModule {}
