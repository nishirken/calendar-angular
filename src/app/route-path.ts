export enum RoutePath {
  Signin = 'auth/signin',
  Signup = 'auth/signup',
  Week = 'week',
}
