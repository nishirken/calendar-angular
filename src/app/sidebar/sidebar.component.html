<wired-calendar #datepicker [attr.data-test-id]="testIds.datepicker"></wired-calendar>