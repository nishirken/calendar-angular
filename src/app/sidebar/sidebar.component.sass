:host
    display: flex
    flex-direction: column
    box-sizing: border-box
    width: 330px
    padding: 35px 20px
    height: 100%
    