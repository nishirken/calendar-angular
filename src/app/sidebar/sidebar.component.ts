import { Component, ElementRef, AfterViewInit, ViewChild } from '@angular/core';
import { Router } from '@angular/router';
import { format } from 'date-fns';
import { DateTimeService } from '../services/date-time/date-time.service';
import { routePaths } from '../routing/routes-paths';
import { sidebarTestIds } from 'calendar-test-ids';

const months = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
];

type CalendarDateSelectEvent = CustomEvent<{
  selected: string;
}>;

@Component({
  selector: 'app-sidebar',
  templateUrl: './sidebar.component.html',
  styleUrls: ['./sidebar.component.sass'],
})
export class SidebarComponent implements AfterViewInit {
  @ViewChild('datepicker') datepicker: ElementRef<any> | null = null;
  testIds = sidebarTestIds;

  constructor(
    private readonly dateTimeService: DateTimeService,
    private readonly router: Router
  ) {}

  ngAfterViewInit(): void {
    if (!this.datepicker) {
      throw new Error('No datepicker has been found');
    }

    this.datepicker.nativeElement.addEventListener(
      'selected',
      this.handleDateSelect
    );

    this.dateTimeService.selectedDate$.subscribe((date) => {
      if (this.datepicker) {
        this.datepicker.nativeElement.value = {
          text: format(date, 'MMM d, yyyy'),
          date,
        };
      }
    });
  }

  handleDateSelect = (event: Event) => {
    event.preventDefault();
    const { detail } = event as CalendarDateSelectEvent;
    const [dayMonth, year] = detail.selected.split(',');
    const [monthName, day] = dayMonth.split(' ');
    this.router.navigateByUrl(
      routePaths.week(
        new Date(
          Number(year),
          months.findIndex((m) => m === monthName),
          Number(day)
        )
      )
    );
  };
}
