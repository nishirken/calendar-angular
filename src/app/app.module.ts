import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';

import { AppRoutingModule } from './app-routing.module';
import { AppComponent } from './app.component';
import { AuthModule } from './auth/auth.module';
import { ErrorCatchingInterceptor } from './services/error-interceptor';

import 'wired-elements/lib/wired-button';
import 'wired-elements/lib/wired-input';
import 'wired-elements/lib/wired-link';

@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule, HttpClientModule, AppRoutingModule, AuthModule],
  providers: [
    {
      provide: HTTP_INTERCEPTORS,
      useClass: ErrorCatchingInterceptor,
      multi: true,
    },
  ],
  bootstrap: [AppComponent],
})
export class AppModule {}
