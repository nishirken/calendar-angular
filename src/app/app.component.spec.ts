import { HttpClientModule } from '@angular/common/http';
import { TestBed } from '@angular/core/testing';
import { BrowserModule } from '@angular/platform-browser';
import { AppComponent } from './app.component';
import { AuthModule } from './auth/auth.module';
import { AppRoutingModule } from './routing/app-routing.module';
import { SidebarModule } from './sidebar/sidebar.module';
import { WeekModule } from './week/week.module';

describe.skip('AppComponent', () => {
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        BrowserModule,
        HttpClientModule,
        AppRoutingModule,
        AuthModule,
        SidebarModule,
        WeekModule,
      ],
      declarations: [AppComponent],
    }).compileComponents();
  });

  it('Renders', () => {
    const fixture = TestBed.createComponent(AppComponent);
    expect(fixture).toMatchSnapshot();
  });
});
