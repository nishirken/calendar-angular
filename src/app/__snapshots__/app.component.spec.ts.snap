// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent Renders 1`] = `
<app-root
  __ngContext__="0"
>
  <router-outlet />
</app-root>
`;
